### ISSUE TITLE
Order of scripts and command

### ISSUE THREAD

**Issue by lionskape** [evilmartians/lefthook#132]
Are there any way to control the sequence of scripts and commands?

For example to execute scripts after commands?
Or to execute the first command, after that execute the first script, and after that execute the second command?

**Comment by Arkweid** [comment]
Only with this way: [piped-option](https://github.com/Arkweid/lefthook/blob/master/docs/full_guide.md#piped-option)
Userful for migrations and similar stuff.

**Comment by lionskape** [comment]
Is it possible to control order between commands and scripts?
Like - command_1, script_1, command_2

Piped example shows only order between commands.

**Comment by Arkweid** [comment]
No, only commands.

**Comment by sanmai-NL** [comment]
New location of docs: https://github.com/evilmartians/lefthook/blob/master/docs/configuration.md#piped

**Comment by BnGx** [comment]
Currently the hooks are executed in alphabetical order, however one of the following options would instead be appropriate:

- Leave the definition order unchanged (including `interactive` hooks and those extended with the `extends` option)
- Insert a specific option (`priority` or `order`) that has a default value (for example 50) and that it can be changed in order to modify the order of execution.

```yaml
database:
  piped: true
  commands:
    create:
      priority: 40
      run: rake db:create
    migrate:
      run: rake db:migrate
    seed:
      priority: 60
      run: rake db:seed
```

This would improve the understanding of the lefthook configuration without introducing distortions.

**Comment by mrexox** [comment]
@BnGx , interactive hooks always run after non-interactive hooks. Scripts run after the all commands. The order is alphanumeric, so you can do something like

```yml
hook:
  commands:
    1_do_this:
      run: ...
    2_then_this:
      run: ...
    ...
```

It probably makes sense to allow a special order of scripts + commands, but why? Would you describe a case when this is really needed and alphanumeric order is not enough?

**Comment by BnGx** [comment]
As I see it, implementing a priority option in Lefthook would be a better approach for several reasons:

- **Enhanced Flexibility and Precision**: A priority-based system would enable developers to accurately determine the execution sequence of hooks, which is vital in complex projects where the order can substantially affect both quality and workflow efficiency.
- **Greater Scalability**: The alphabetical method tends to become unwieldy and disorganized with an increasing number of hooks. Conversely, a priority-based approach would remain clear and manageable, even in larger projects.
- **Improved Clarity and Readability**: The alphabetical system often compels developers to use numerical prefixes in hook names, which can detract from code readability and intuitiveness. A priority system, in my opinion, would be more straightforward and self-explanatory.
- **Adaptability to Rapid Changes**: In agile development environments, where requirements can shift swiftly, a priority option would facilitate quick, flexible adjustments in the execution order without necessitating hook renaming or reorganization.
- **Minimization of Errors and Confusion**: Alphabetical ordering can lead to misunderstandings or mistakes, especially for newcomers or during code reviews. I believe a priority system would mitigate these risks by providing a more distinct and structured execution order.

Assuming the configuration is as follows

```yaml
pre-commit:
  piped: true
  commands:
    ...
    15_format_ruby:
      run: ...
    15_format_python:
      run: ...
    20_lint_ruby:
      run: ...
    20_lint_python:
      run: ...
    ...
```

In this case, if I had to execute a hook manually or call it from another tool or in a CI environment I would have to remember the precise name of the hook and write `lefthook run pre-commit --commands 15_format_python,20_lint_python`.

This seems really rather unintuitive to me than to write `lefthook run pre-commit --commands format_python,lint_python` as would happen if a `priority` option was implemented.

In summary, from my perspective, while the alphabetical ordering method may appear simple, it falls short in effectively managing the execution order of hooks in a dynamic and scalable way. A priority-based system, on the other hand, would offer a robust, transparent, and easily manageable solution, aptly suited to the demands of contemporary, complex software projects.

**Comment by sanmai-NL** [comment]
@mrexox By imposing an order by design, commands that are order-dependent cannot always work.
For example, a command may have an auto-fixer and  only after that fixer is done (e.g., syntax migration), other checkers can do their job. Also, some commands make take a longer time (Lefthook can be run in CI too, not just as a quick pre-commit check), and you may want to run them first or last. For example, some data science data validator tools may take longer time than the common source code checkers.

By allowing commands to be a list of dicts, rather than a dict, the user may indicate that order is important in a backward-compatible way.

**Comment by mrexox** [comment]
Ok, I see. So, I guess this is your use case:

```yml
database:
  piped: true
  commands:
    create:
      priority: 40
      run: rake db:create
    review:
      priority: 45
      interactive: true
      run: <some interactive command>
    migrate:
      run: rake db:migrate
    seed:
      priority: 60
      run: rake db:seed
```

And you expect `interactive: true` command to be executed after `create` but before `migrate` and `seed`, right?

**Comment by BnGx** [comment]
Absolutely yes. I believe that in this way, in accordance with what has been said so far, the configuration is more customizable, scalable, simple to maintain and less confusing, especially in large projects.

**Comment by mrexox** [comment]
I think priority of 50 is not obvious as a default. I'd suggest 0 as the default. If you don't specify a priority it is 0, and the command is always executed at the very end.

I think it makes sense since priority should be a natural number.

**Comment by BnGx** [comment]
I specify that I was referring to a decreasing order of priority (0 highest priority, 100 lowest priority), therefore a command with priority 1 is executed before a command with priority 10. According to this consideration, the commands with priority 0 would all be executed for first.

My proposal to use 50 arises from the possibility of being able to integrate the new configurations (those with the `priority` option) with the pre-existing ones, giving the possibility of being able to execute a new command (local or remote) before or after those already existing in the old configurations avoiding having to make further changes to them.

**Comment by sanmai-NL** [comment]
@BnGx What's the value of a `priority` key over an ordered datastructure, like a list?

**Comment by BnGx** [comment]
@mrexox I agree that the integration of the priority option should be explicit. I think the choice of 0 rather than 50 can be addressed simply by customizing the configuration. So if you think it is more appropriate to use 0 for choices related to the programming language in use I think it is right. There is nothing to prevent the operation from being varied based on the future needs of the community, I suppose, with little effort.

@sanmai-NL the introduction of an ordered datastructure is still a valid alternative, but this would involve the introduction of a substantial change that could cause quite a few problems for those who have been using left hooks for some time. Furthermore, there would always be a limitation in correctly extending the configuration using the `extends` and `remote` options.

**Comment by sanmai-NL** [comment]
@BnGx It would not, the data type of commands would be `[dict | list[dict]]` in Python terms. I.e. the union of both types. This means that the dict-based schema is still valid and nothing breaks. What is the problem with `extends` exactly? Merging lists is perfectly valid operation, and it can simply be rejected if multiple lists are specified. When a `priority` is specified, how to merge if the priority values are equal?

**Comment by BnGx** [comment]
Using either dictionaries or a list of dictionaries as a configuration format is acceptable, but, for me, it is advisable to exclusively use one of these formats for the following reasons:

A single format makes the data structure more predictable and easier to understand. When developers or users of your system see a consistent format, they can more easily grasp how the data is organized and how to interact with it. Consistency in data formatting is crucial, especially in complex systems or large projects with many contributors. With a standard format, you reduce the risk of data misinterpretation and facilitate collaboration among team members.

With a uniform configuration format, validation becomes simpler. You can implement consistency checks and data validation more efficiently, reducing the possibility of errors or malformed data. A single, well-defined format is easier to maintain and update. When changes need to be made, it’s simpler to do so in a system with a cohesive data structure rather than one with multiple different formats.

Also consider that the lefthook makes its speed a source of pride so managing a single type of data structure can improve system performance, as algorithms and logic can be optimized for that specific structure.

> (The whole extends feature would be better deprecated and implemented using JSON patch, which is an industry standard to this problem of merging config files in JSON, YAML, TOML. See https://github.com/kubernetes-sigs/kustomize/blob/master/examples/jsonpatch.md for an example in Kubernetes.)

I believe this is a choice of the developers who decided to merge external files using the Yaml library, effectively obtaining a valid and processable YAML file with the tools currently available for formatting and linting.

So the list of dicts or simple dicts is fine, the key is to balance consistency and clarity with the necessary flexibility and functionality, but introducing a new format would mean a greater implementation and maintenance effort.

**Comment by sanmai-NL** [comment]
@BnGx I think the arguments you make don't weigh up against breaking backward compatibility. As a middle ground, the dictionary-based `commands` can be deprecated in favor of the list of dictionaries-based `commands`. Also, a version field can be put inside the Lefthook YAML to give the maintainers more freedom in improving the format.

**Comment by BnGx** [comment]
So do you propose to abandon dictionaries in the future? How would you make existing old configurations compatible?

**Comment by sanmai-NL** [comment]
Older configurations can be run by an older version of Lefthook. But only if you deprecate the current format, which is one of your suggestions.

**Comment by mrexox** [comment]
I am not planning to change the format of commands. One of the key features of lefthook is the ability to use `lefthook.yml` and a `lefthook-local.yml` with some overrides. With dictionary approach this is done with no effort via YAML processing library. And I don't yet understand the value of a list structure, since (as I see) the results can be reached with the existing options.

By the way, `priority` option has just been release with **1.5.5** version. Only for commands now.

**Comment by mrexox** [comment]
Right now there's no way to run some commands sequential and some in parallel. Only if using sub-hooks and running them from within the another command, like in the [example](https://github.com/evilmartians/lefthook/issues/132#issuecomment-1833411442) I provided before. This is not a perfect solution and there might be issues with `remote` and `extends` options too. But if you describe a real use case, I will try to figure out what needs to be improved in lefthook configuration.

I agree that having commands with the same priority (or configured in a list) go parallel and others go sequentially is convenient. But I really need to understand a use case and that there's no other way to do that, rather than change lefthook configuration format.

**Comment by mrexox** [comment]
Guys! What do you think about nested configuration, something like this:

```yml
pre-commit:
  parallel: true
  do:
    - run: yarn lint {staged_files}
    - run: yarn test {staged_files}
    - script: my-custom-script.sh
      runner: bash
    - name: a piped group
      group:
        piped: true
        do:
          - run: yarn step-1
          - run: yarn step-2
          - script: my-custom-js-script.js
```

I believe this would solve everyone's problems, but I think it might take some time to implement properly. This won't interfere with the current configuration however adds much more flexibility in a way you can group the parallel/piped scripts and commands. WDYT? Does it sound like a solution? 

Of course, it's just an idea, so any suggestions to naming for **do** and **group** parts which are new here – are welcome
